core/src/main/java/org/apache/iceberg/deletes/Deletes.java

```diff
@@ -137,93 +150,91 @@ protected boolean shouldKeep(T row) {
+      this.deletePosIterator = deletePositions.iterator();
     }
 
     @Override
     public CloseableIterator<T> iterator() {
-      CloseableIterator<Long> deletePosIterator = deletePositions.iterator();
-
       CloseableIterator<T> iter;
       if (deletePosIterator.hasNext()) {
-        iter = new PositionFilterIterator(rows.iterator(), deletePosIterator);
+        nextDeletePos = deletePosIterator.next();
+        iter = createPosDeleteIterator(rows.iterator());
       } else {
         iter = rows.iterator();
-        try {
-          deletePosIterator.close();
-        } catch (IOException e) {
-          throw new UncheckedIOException("Failed to close delete positions iterator", e);
-        }
       }
 
       addCloseable(iter);
+      addCloseable(deletePosIterator);
 
       return iter;
     }
 
-    private class PositionFilterIterator extends FilterIterator<T> {
-      private final CloseableIterator<Long> deletePosIterator;
-      private long nextDeletePos;
+    boolean isDeleted(T row) {
+      long currentPos = rowToPosition.apply(row);
+      if (currentPos < nextDeletePos) {
+        return false;
+      }
 
-      protected PositionFilterIterator(CloseableIterator<T> items, CloseableIterator<Long> deletePositions) {
-        super(items);
-        this.deletePosIterator = deletePositions;
+      // consume delete positions until the next is past the current position
+      boolean isDeleted = currentPos == nextDeletePos;
+      while (deletePosIterator.hasNext() && nextDeletePos <= currentPos) {
         this.nextDeletePos = deletePosIterator.next();
+        if (!isDeleted && currentPos == nextDeletePos) {
+          // if any delete position matches the current position
+          isDeleted = true;
+        }
       }
 
-      @Override
-      protected boolean shouldKeep(T row) {
-        long currentPos = extractPos.apply(row);
-        if (currentPos < nextDeletePos) {
-          return true;
-        }
+      return isDeleted;
+    }
+
+    protected abstract CloseableIterator createPosDeleteIterator(CloseableIterator<T> items);
+  }
+
+  private static class PositionStreamDeleteFilter<T> extends PositionStreamDeleteIterable<T> {
+    private PositionStreamDeleteFilter(CloseableIterable<T> rows, Function<T, Long> rowToPosition,
+                                       CloseableIterable<Long> deletePositions) {
+      super(rows, rowToPosition, deletePositions);
+    }
 
-        // consume delete positions until the next is past the current position
-        boolean keep = currentPos != nextDeletePos;
-        while (deletePosIterator.hasNext() && nextDeletePos <= currentPos) {
-          this.nextDeletePos = deletePosIterator.next();
-          if (keep && currentPos == nextDeletePos) {
-            // if any delete position matches the current position, discard
-            keep = false;
-          }
+    @Override
+    protected CloseableIterator createPosDeleteIterator(CloseableIterator<T> items) {
```

Comment:
nit: raw type

Reply:
I'd say it's a bit more than a nit! Iceberg PRs should always use the type system.

Reply:
Fixed it in a new commit.